### PR TITLE
[FEATURE] Cosmos | Connection endpoint rediscovery (connection state listener)

### PR DESCRIPTION
**Pull request by David-Noble-at-work**
#### TODO
- [x] Feature description
- [x] Read latency benchmarks
       Collected and analyzed numbers from our prod-03 environment
       See [read latency benchmark results](#read-latency-benchmark-results)
- [X] Socket utilization numbers
       Collected numbers from our prod-03 environment (As part of read latency benchmark run)
       See [read latency benchmark results](#read-latency-benchmark-results)
- [X] Memory utilization numbers
      Collected numbers from our prod-03 environment (As part of read latency benchmark run)
      See [read latency benchmark results](#read-latency-benchmark-results)
- [X] CPU utilization numbers
      Collected numbers from our prod-03 environment (As part of read latency benchmark run)
      See [read latency benchmark results](#read-latency-benchmark-results)
- [x] Thought experiment that builds confidence that this feature corrects other read latency issues we've been seeing
       Reviewed with @xinlian12, @kushagraThapar, @moderakh
- [ ] Upgrade test results with evidence that this feature functions as expected without performance or reliability issues.
       Collecting data from our test-33 environment
       Next up: Analyze results

This PR will be ready for merge when we're convinced that the feature is functionally correct and does not introduce performance or reliability issues.

#### Purpose

The connection endpoint rediscovery feature is designed to reduce and spread-out high latency spikes that are likely to occur:

- During rolling upgrades of a Cosmos instance or
- When a backend node is being decommissioned or restarted (e.g., to restart or remove an unhealthy replica)

Our expectation is that this PR will improve reliability without effecting performance under normal circumstances when there is no rolling upgrade taking place and no servers being decommissioned or restarted.

#### Implementation

The `GlobalAddressResolver` now implements `AddressResolverExtension`. This new interface extends `IAddressResolver` and adds methods to support the `RntbdConnectionStateListener`. The `RntbdTransportClient` now uses the `RntbdConnectionStateListener` to 
build a reverse lookup table that maps physical endpoint addresses to sets of `PartitionKeyRangeIdentity` instances. New endpoint addresses are added each time that `RntbdTransportClient::invokeStoreAsync` is called.

When we detect that a server may be down or going down, we remove the effected `PartitionKeyRangeIdentity` instances from the physical address cache and close the effected `RntbdEndpoint`. The physical addresses that service an effected `PartitionKeyRangeIdentity` will then be updated on the next request as if it were the first request targeting the `PartitionKeyRangeIdentity`. Here are the conditions that `RntbdTransportClient::invokeStoreAsync` detects:

- A graceful shutdown is occurring.

  This is indicated by a `GoneException` with sub-status code zero (`SubStatusCodes.UNKOWN`).

- The service is down.

  This is indicated by a `GoneException` with non-null cause. We expect an `IOException` as the cause. Common causes of type `IOException` are:

  - `ConnectTimeoutException`

    Indicates that an attempt to connect to a remote host timed out.
  
  - `ConnectException`

    Indicates that the server can't be reached.
  
  - `ClosedChannelException`

    Indicates a connection dropped.

  - `IOException`

    Indicates a connection was reset by the remote host.

Endpoint closure is aggressive. We close an endpoint (all channels) at the time that a connection issue is detected on any channel. We will consider less aggressive strategies based on what we find in test. We are cognizant of the fact that unnecessary `RntbdEndpoint` evictions could cause a flood of retries. We will adapt to what we find in test.

This feature is hidden behind a new feature flag: `DirectConnectionConfig::enableConnectionEndpointRediscovery` (default: false). When we're satisfied with the implementation, we'll change the default to `true`.

#### Read Latency Benchmark results

All benchmark results were produced in the prod-03 test environment:

* Cosmos DB instance:

  cosmos-sdk-core-3

* Virtual machine:

  Linux (ubuntu 18.04)

  Standard F16s_v2 (16 vcpus, 32 GiB memory)

[Here](https://microsoft-my.sharepoint.com/:u:/p/danoble/EYEa8SnFIkdIvNHnwS7FwwUBMJwx6VYoDFRV79V6kmlx9w?e=veDu6X) is the raw data and a Jupyter notebook for further analysis.

##### Charts

1. P95, Document count: 1

image

2. P95, Document count: 100,000

latency-p95 documentCount=100000

3. P99, Document count: 1

latency-p99 documentCount=1

4. P99, Document count: 100,000

latency-p99 documentCount=100000

5. P99.9, Document count: 1

latency-p999 documentCount=1

6. P99.9, Document count: 100,000

latency-p999 documentCount=100000

7. Throughput, Document count: 1

throughput documentCount=1

8. Throughput, Document count: 100,000

throughput documentCount=100000

9. Socket utilization, Document count: 1

This chart shows socket counts by concurrency level with `documentCount` `1`. Socket counts are summed over the set of observations taken at the end of each of five one minute periods in the benchmark. Socket counts were collected using ss. Divide the socket count by five to get the average count at the end of each period.

socket_utilization documentCount=000001

10. Socket utilization, Document count: 100,000

This chart shows socket counts by concurrency level with `documentCount` `100,000`. Socket counts are summed over the set of observations taken at the end of each of five one minute periods in the benchmark. Socket counts were collected using ss. Divide the socket count by five to get the average count at the end of each period.

socket_utilization documentCount=100000

11. Virtual memory utilization, Document count: 1

This chart shows virtual memory utilization in mebibytes by concurrency level with `documentCount` `1`. Results are averaged over the set of observations taken at the end of each of five one minute periods in the benchmark. Virtual memory numbers were collected using top.

virtual_memory documentCount=000001

12. Virtual memory utilization, Document count: 100,000

This chart shows virtual memory utilization in mebibytes by concurrency level with `documentCount` `100,000`. Results are averaged over the set of observations taken at the end of each of five one minute periods in the benchmark. Virtual memory numbers were collected using top.

virtual_memory documentCount=100000

13. CPU utilization, Document count: 1

This chart shows CPU utilization as a percentage by concurrency level with `documentCount` `1`. Results are averaged over the set of observations taken at the end of each of five one minute periods in the benchmark. CPU utilization numbers were collected using top.

cpu documentCount=000001

14. CPU utilization, Document count: 100,000

This chart shows CPU utilization as a percentage by concurrency level with `documentCount` `100,000`. Results are averaged over the set of observations taken at the end of each of five one minute periods in the benchmark. CPU utilization numbers were collected using top.

cpu documentCount=100000

#### Later: In a follow-on PR

- Distinguish between requests that fail because of a connection issue:
  
  - Writes (aka, sends) that do not reach the server
  - Writes (aka, sends) that reach the server (i.e., confirmed by TCP ACK)

  Rationale: enables retries for create/delete/update operations that are known not to reach the server. `TransportException` was developed with this in mind. This info can be extracted from the `RequestTimeline` and stored into the exception.

- Ensure that Cosmos client never instantiates `GoneException` with sub-status code zero.

  Consider whether the client should instantiate any `CosmosException` using a sub-status code that may be returned in a response from the server.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GlobalAddressResolver.java [hunk 1]
```diff
@@ -77,46 +80,87 @@ public GlobalAddressResolver(
...
+    @Override
+    public URI getAddressResolverURI(RxDocumentServiceRequest rxDocumentServiceRequest) {
+        return this.endpointManager.resolveServiceEndpoint(rxDocumentServiceRequest);
+    }
+
+    @Override
+    public boolean removeAddressResolverURI(RxDocumentServiceRequest request) {
+        final URI uri = this.getAddressResolverURI(request);
+        this.addressCacheByEndpoint.computeIfPresent(uri, (address, cache) -> {
+            // TODO (DANOBLE) What work must be done here?
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GlobalAddressResolver.java:98]: as discussed on the code review session. here you need to do a surgery in GatewayAddressCache to remove the corresponding stale replica address.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.azure.cosmos.implementation.directconnectivity;
+
+import com.azure.cosmos.implementation.RxDocumentServiceRequest;
+import com.azure.cosmos.implementation.directconnectivity.rntbd.RntbdAddressCacheToken;
+import reactor.core.publisher.Mono;
+
+import java.net.URI;
+import java.util.List;
+
+public interface AddressResolverExtension extends IAddressResolver {
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java:10]: please document the purpose of these APIs.

**Review by moderakh — CHANGES_REQUESTED** [review]
See the description which now includes the substance of this comment.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GlobalAddressResolver.java (hunk 1 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GlobalAddressResolver.java:98]: Please fill the implementation gap.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 3]
```diff
@@ -79,6 +81,28 @@ public DirectConnectionConfig setConnectTimeout(Duration connectTimeout) {
         return this;
     }
 
+    /**
+     * Gets a value that indicates whether Direct TCP connection endpoint rediscovery should is enabled.
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:85]: Expand on what it is from CX perspective. 

QQ; Do we want this feature flag as prominent in ConnectionConfig?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java [hunk 4]
```diff
@@ -288,25 +288,32 @@
...
         public static final int DATABASE_ACCOUNT_NOTFOUND = 1008;
 
         // 404: LSN in session token is higher
         public static final int READ_SESSION_NOT_AVAILABLE = 1002;
+
+        // 410: Gone sub-status
+        // Sub-status code zero in a response from a service endpoint indicates that a replica is being discontinued or
+        // reconfigured. When endpoint rediscovery is enabled the RntbdTransportClient converts sub-status code zero to
+        // this sub-status code value.
+        public static final int DISCONTINUING_SERVICE = 10002;
```
[the remark this answers is not in the document]
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java]: Lets reserve some status code ranges in master

---
In reply to: [481464735](https://github.com/Azure/azure-sdk-for-java/pull/14697#discussion_r481464735) [](ancestors = 481464735)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java [hunk 5]
```diff
@@ -288,25 +288,37 @@
...
+        // 410: Gone sub-status
+        // Sub-status code indicating that a GoneException was instantiated by the client; not as a result of a response
+        // message provide by a Cosmos instance.
+        public static final int CLIENT_GENERATED = 10_000;
+
+        // 410: Gone sub-status
+        // Sub-status code zero in a response from a service endpoint indicates that a replica is being discontinued or
+        // reconfigured. When endpoint rediscovery is enabled the RntbdTransportClient converts sub-status code zero to
+        // this sub-status code value.
+        public static final int DISCONTINUING_SERVICE = CLIENT_GENERATED + 2;
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:321]: How doe these sub-status code impact GoneRetryPolicy?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java [hunk 6]
```diff
@@ -172,11 +169,12 @@ public Event(final String name, final Instant from, final Instant to) {
             this.name = name;
             this.startTime = from;
 
-            this.duration = from == null ? null : to == null ? Duration.ZERO : Duration.between(from, to);
-            if(this.duration != null) {
-                this.durationInMicroSec = duration.toNanos()/1000L;
+            if (from == null) {
+                this.duration = null;
+            } else if (to == null) {
+                this.duration = Duration.ZERO;
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java:175]: Right interpretation of NULL and ZERO is hard. 
How about populate Duration only when end-time is present?

Or even better is list Start and EndTimes

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    URI getAddressResolverURI(RxDocumentServiceRequest request);
+    boolean removeAddressResolverURI(RxDocumentServiceRequest request);
+    Mono<Void> updateAsync(RxDocumentServiceRequest request, List<RntbdAddressCacheToken> tokens);
```
[the remark this answers is not in the document]
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java]: Am for right code even if its mean deviating from .NET code.

---
In reply to: [481473446](https://github.com/Azure/azure-sdk-for-java/pull/14697#discussion_r481473446) [](ancestors = 481473446)

[the remark this answers is not in the document]
Comment (kirankumarkolli) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java]: Also how about inline this into the AddresResolver interface?

.NET extension model is language specific, may be we can inline in Java?

---
In reply to: [482776313](https://github.com/Azure/azure-sdk-for-java/pull/14697#discussion_r482776313) [](ancestors = 482776313,481473446)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java [hunk 8]
```diff
@@ -135,6 +137,16 @@ public GatewayAddressCache(
              DefaultSuboptimalPartitionForceRefreshIntervalInSeconds);
     }
 
+
+    @Override
+    public void removeAddresses(final PartitionKeyRangeIdentity partitionKeyRangeIdentity) {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java:142]: Is remove un-conditional or conditional?

I.e. in-case of high throughut clients simple remove might result in removing existing valid refresh entries. 
Do they need to be based on the stale state instead?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java [hunk 9]
```diff
@@ -78,13 +79,17 @@ public static RntbdRequest decode(final ByteBuf in) {
     void encode(final ByteBuf out) {
 
         final int expectedLength = RntbdRequestFrame.LENGTH + this.headers.computeLength();
-        final int start = out.readerIndex();
+        checkState(out.readerIndex() == out.writerIndex());
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java:82]: Suggestion: Can unrelated changes be  pushed as separate iteration?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java [hunk 10]
```diff
@@ -487,7 +487,7 @@ public void write(final ChannelHandlerContext context, final Object message, fin
             final RntbdRequestRecord record = (RntbdRequestRecord) message;
             this.timestamps.channelWriteAttempted();
 
-            context.write(this.addPendingRequestRecord(context, record), promise).addListener(completed -> {
+            context.writeAndFlush(this.addPendingRequestRecord(context, record), promise).addListener(completed -> {
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:490]: This was something I have asked before and was told is not needed.
Can you please expand on why is it needed and how is it related to ConnectionStateListener.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java [hunk 11]
```diff
@@ -65,18 +65,18 @@
         new ClosedChannelException(), RntbdClientChannelPool.class, "acquire");
 
     private static final IllegalStateException POOL_CLOSED_ON_ACQUIRE = ThrowableUtil.unknownStackTrace(
-        new StacklessIllegalStateException("service endpoint was closed while acquiring a channel"),
+        new ChannelAcquisitionException("service endpoint was closed while acquiring a channel"),
         RntbdClientChannelPool.class, "acquire");
 
     private static final IllegalStateException POOL_CLOSED_ON_RELEASE = ThrowableUtil.unknownStackTrace(
-        new StacklessIllegalStateException("service endpoint was closed while releasing a channel"),
+        new ChannelAcquisitionException("service endpoint was closed while releasing a channel"),
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:72]: Is release also related to Acuisition?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java [hunk 12]
```diff
@@ -65,18 +65,18 @@
...
         RntbdClientChannelPool.class, "release");
 
     private static final AttributeKey<RntbdClientChannelPool> POOL_KEY = AttributeKey.newInstance(
         RntbdClientChannelPool.class.getName());
 
     private static final IllegalStateException TOO_MANY_PENDING_ACQUISITIONS = ThrowableUtil.unknownStackTrace(
-        new StacklessIllegalStateException("too many outstanding acquire operations"),
+        new ChannelAcquisitionException("too many outstanding channel acquisition operations"),
```
Comment (kirankumarkolli) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:79]: Will the exception type change impact any upstream error/exception handling semantics.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 13]
```diff
@@ -79,6 +81,28 @@ public DirectConnectionConfig setConnectTimeout(Duration connectTimeout) {
...
+     *
+     * @return {@code true} if Direct TCP connection endpoint rediscovery should is enabled; {@code false} otherwise.\
+     */
+    public boolean getEnableConnectionEndpointRediscovery() {
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:89]: good catch. thanks.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java [hunk 14]
```diff
@@ -79,6 +81,28 @@ public DirectConnectionConfig setConnectTimeout(Duration connectTimeout) {
...
+
+    /**
+     * Sets a value that indicates whether Direct TCP connection endpoint rediscovery should is enabled.
+     *
+     * @param enableConnectionEndpointRediscovery {@code true} if connection endpoint rediscovery is enabled; {@code
+     *                                            false} otherwise.
+     *
+     * @return the {@linkplain DirectConnectionConfig}.
+     */
+    public DirectConnectionConfig setEnableConnectionEndpointRediscovery(boolean enableConnectionEndpointRediscovery) {
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:101]: good catch. thanks.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java [hunk 15]
```diff
@@ -493,7 +515,7 @@ public String toString() {
             ", inetSocketProxyAddress=" + (proxy != null ? proxy.getAddress() : null) +
             ", readRequestsFallbackEnabled=" + readRequestsFallbackEnabled +
             ", connectTimeout=" + connectTimeout +
-            ", idleEndpointTimeout=" + idleEndpointTimeout +
+            ", idleEndpointTimeout=" + idleTcpEndpointTimeout +
             ", maxConnectionsPerEndpoint=" + maxConnectionsPerEndpoint +
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java:519]: yep. once again, good catch. thanks.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java]: Agreed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java [hunk 16]
```diff
@@ -156,13 +156,10 @@ public String toString() {
         @JsonIgnore
         private final Duration duration;
 
-        @JsonSerialize(using = ToStringSerializer.class)
-        private final long durationInMicroSec;
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java:160]: It is reported, just not stored as a value. Are you saying that it is useful for debugging?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
+package com.azure.cosmos.implementation.directconnectivity;
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java:1]: thanks. change made in local branch

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:101]: change made in local branch

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java]: We have two implementations of the IAddressResolver interface: AddressResolver and GlobalAddressResolver. AddressResolver implements IAddressResolver and GlobalAddressResolver implements AddressResolverExtension. This is as .NET has done it.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java (hunk 2 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/AddressResolverExtension.java:10]: I've added a minimal description to the code on my local branch. We might elaborate and also document the IAddressResolver interface in follow-on PR.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java [hunk 18]
```diff
@@ -41,16 +41,18 @@
...
     private int maxConnectionsPerEndpoint;
     private int maxRequestsPerConnection;
-    private Duration idleTcpConnectionTimeout;
 
     /**
      * Constructor.
      */
     public ConnectionPolicy(GatewayConnectionConfig gatewayConnectionConfig) {
         this(ConnectionMode.GATEWAY);
+        this.enableTcpConnectionEndpointRediscovery = false;
```
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ConnectionPolicy.java:55]: Of course. It is set as it is at your request. I'm good with making the default true. The change is on my branch.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java:89]: change made in local branch.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java (hunk 12 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:79]: No, it will not.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java (hunk 11 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdClientChannelPool.java:72]: It's what happens immediately after a channel is acquired and the request is dispatched to the channel. We've got a promise that the write will occur, but have not yet received a TCP ACK.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java (hunk 10 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:490]: Agreed. Thanks for catching that. It is not required and the change was likely the result of an experiment that I was running. I flush further down the pipeline.

Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequestManager.java:490]: change is on my working branch

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java (hunk 9 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java:82]: This is a change that I'll keep in. It is a potential bug that I caught while debugging. I must check the writerIndex, not the reader index. It is luck that we haven't had a problem here.

Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java:82]: I did remove the checkState line as this was a check to ensure I hadn't broken anything based on correcting the code. This verified that I wasn't depending on the bug.

Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java:82]: change is in my branch

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java (hunk 8 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/GatewayAddressCache.java:142]: This is very much worth discussion. I'll give this one some thought.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java (hunk 6 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java:175]: I'll consider this change after this PR.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java (hunk 5 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:321]: Confirmed by way of the logs on test33 and also by code inspection. This does not effect retry policy.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java (hunk 9 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/RntbdRequest.java:82]: change committed and pushed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 3 above)
Comment (David-Noble-at-work) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java]: It is a good question whether we want this flag as prominent as it is. This PR now enables connection endpoint rediscovery by default. I would prefer not to advertise the feature until we've got more experience with it. Enabling the feature by default is counter to that.

@kushagraThapar, @moderakh what do you think?

**Review by moderakh — CHANGES_REQUESTED** [review]
There are code style change in this PR. 

the direct stack is complex, with code style change it is not easy to review the change.

Could you revert any code style change please? that should help in validating the code in the code review

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java (hunk 3 above)
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/DirectConnectionConfig.java]: IMO, any feature may have bug on its first release. I prefer if we change the default to enabled later when we are more confident on the behaviour.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RequestTimeline.java:160]: The value will still be logged:
"serializationDiagnosticsContext":{"serializationDiagnosticsList":[{"serializationType":"ITEM_SERIALIZATION","startTimeUTC":"22 Sep 2020 17:40:59.239","endTimeUTC":"22 Sep 2020 17:40:59.239","durationInMicroSec":0},{"serializationType":"PARTITION_KEY_FETCH_SERIALIZATION","startTimeUTC":"22 Sep 2020 17:40:59.317","endTimeUTC":"22 Sep 2020 17:40:59.323","durationInMicroSec":5998}]},"gatewayStatistics":null,"systemInformation":{"usedMemory":"60586 KB","availableMemory":"8309590 KB","systemCpuLoad":"(2020-09-22T17:40:58.729927700Z 20.1%)"}}

**Comment by xinlian12** [comment]
This PR has been split into three PRs:
https://github.com/Azure/azure-sdk-for-java/pull/16204
https://github.com/Azure/azure-sdk-for-java/pull/16197
https://github.com/Azure/azure-sdk-for-java/pull/15991
